Clear Containers for Docker Engine (v1.8.3)

jamesodhunt opened the pull request:
This is a follow-on to the original Clear Containers for Docker PR #16163 .

The branch still uses the "clr" execution driver since docker has not yet migrated to calling out to runc (OCI).

Commit:
Clear Containers for Docker Engine execution driver

Signed-off-by: Dimitri John Ledkov <dimitri.j.ledkov@intel.com>
Signed-off-by: James Hunt <james.o.hunt@intel.com>
Signed-off-by: Michael Doherty <michael.i.doherty@intel.com>

vbatts committed:
devicemapper: fix zero-sized field access

Fixes: #15279

Due to
https://github.com/golang/go/commit/7904946eeb35faece61bbf6f5b3cc8be2f519c17
the devices field is dropped.

This solution works on go1.4 and go1.5

Signed-off-by: Vincent Batts <vbatts@redhat.com>

Commit:
Merge branch 'release/1.8' of git://github.com/docker/docker into new-merge

Signed-off-by: Dimitri John Ledkov <dimitri.j.ledkov@intel.com>

Commit:
Properly find the right Endpoint to steal networking information from.

Signed-off-by: Dimitri John Ledkov <dimitri.j.ledkov@intel.com>

Commit:
Fix ups.

Signed-off-by: Dimitri John Ledkov <dimitri.j.ledkov@intel.com>

Commit:
Change version tag

Signed-off-by: Dimitri John Ledkov <dimitri.j.ledkov@intel.com>

Commit:
Merge tag 'v1.8.3' into github-clr-1.8.3-submit

v1.8.3

Signed-off-by: James Hunt <james.o.hunt@intel.com>

Conflicts:
	VERSION

Commit:
Change version tag.

Signed-off-by: James Hunt <james.o.hunt@intel.com>

runcom commented:
@jamesodhunt is this PR for "educational & testing purpose" as the old one? (cit. https://github.com/docker/docker/pull/16163#issuecomment-139024189)

runcom commented on pkg/parsers/operatingsystem/operatingsystem_linux.go:
```diff
@@ -12,12 +13,22 @@ var (
 	etcOsRelease = "/etc/os-release"
+
+	// used by stateless systems like Clear Linux
+	altEtcOSRelease = "/usr/lib/os-release"
```
see #17174 hope to get this merged to fix this :)

runcom commented:
is this also supposed to work on non-clear-linux host? cause I get:
```
Error response from daemon: Cannot start container 287c38642d38f1b6bc00bd1f319068d234ff325454b75b7ae14e13a697243e46: fork/exec lkvm: no such file or directory
```
(following the ML I came across this last month https://lists.clearlinux.org/pipermail/dev/2015-September/000042.html)

runcom committed:
Dockerfile: add kvmtool and linux-container for clr execdriver

This patch adds the needed pkgs to docker/docker development Dockerfile
in order to test the clr execdriver in the development container.

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

jamesodhunt commented:
Hi @runcom - since docker is not yet calling out to `runc`, we would like you to consider merging this branch as this is the only way to use an alternative runtime at present.

Note that we also have a basic OCI-compatible runtime for Clear Containers at:

- https://github.com/clearlinux/kvmtool/tree/oci.

Once docker starts calling an OCI runtime directly, we would be able to switch between the default and our runtime, using something like:

```
docker daemon --exec-opt "oci.runtime.cmd=lkvm" ...
```

At that point, this exec-driver could be dropped. However, it is unclear when docker will switch to OCI (docker 2.0?)

runcom commented:
@jamesodhunt thanks for the info! Do you actually ship https://github.com/clearlinux/kvmtool/tree/oci with Clear Linux (`lkvm`) or are you using the original kvmtool from kernel.org?

jamesodhunt commented:
@runcom - we are currently shipping the Clear Containers exec driver, and are awaiting an OCI-capable docker to validate our OCI-lkvm work before shipping it as stable.

runcom commented:
I'm getting `DEBU[0504] Remove interface failed: Link not found` in daemon logs when doing a simple `docker run busybox echo "hello world" (Fedora 22)

runcom commented:
`docker run busybox /bin/sh` gives you a shell prompt but the hangs also
